sesman/chansrv/chansrv_config.c

```diff
@@ -127,7 +128,47 @@ read_config_security(log_func_t logmsg,
 
         if (g_strcasecmp(name, "RestrictOutboundClipboard") == 0)
         {
-            cfg->restrict_outbound_clipboard = g_text2bool(value);
+            if (g_text2bool(value))
+            {
+                cfg->restrict_outbound_clipboard = 1;
+            }
+            else
+            {
+                if (g_find_property2bool(value, "text"))
+                {
+                    cfg->restrict_outbound_clipboard_text = 1;
+                }
+                if (g_find_property2bool(value, "file"))
+                {
+                    cfg->restrict_outbound_clipboard_file = 1;
+                }
+                if (g_find_property2bool(value, "image"))
+                {
+                    cfg->restrict_outbound_clipboard_image = 1;
+                }
```

Comment:
This approach is simple and readable, but it doesn't address a current problem with the existing code which is that a mis-spelled keyword is not detected. Also, this part of the code can't easily be tested.

Like @metalefty, I think you want to use a bitmask to solve this, and an array of keywords and values which compute your value.

We have a structure in string calls.h called `struct bitmask_string` and a function called `g_bitmask_to_str()`. The function is a general way to convert a bitmask value into a string for logging.

What we could do with is a function called `g_str_to_bitmask()` which works the other way. Here's a signature:-

```c
/***
 * Converts a string containing a series of tokens to a bitmask.
 * @param str Input string
 * @param bitmask_string Array mapping tokens to bitmask values
 * @param delim Delimiter for tokens in str
 * @param[out] unrecognised Buffer for any unrecognised tokens
 * @param unrecognised_len Length of unrecognised including '`0';
 * @return bitmask value for recognised tokens
 */
int
g_str_to_bitmask(const char *str, const struct bitmask_string[],
                 char delim, char *unrecognised,
                 int unrecognised_len);
```

If the function encounters an unexpected token, it's added to unrecognised with a delimiter if necessary, which could be the passed-in delimiter or a space or whatever.

Then in chansrv_config.h you could have something like:-

```c
/* Define bitmask values for restricting the clipboard */
#define CLIP_RESTRICT_TEXT (1<<0)
#define CLIP_RESTRICT_FILE (1<<1)
#define CLIP_RESTRICT_IMAGE (1<<2)
#define CLIP_RESTRICT_ALL 0xffffffff
```

and in chansrv_config.c, rather than the above you could have:-

```c
/* Map clipboard strings into bitmask values */
const struct bitmask_string restrict_map[] = 
{
    { CLIP_RESTRICT_TEXT, "text"},
    { CLIP_RESTRICT_FILE, "file"},
    { CLIP_RESTRICT_IMAGE, "image"},
    { CLIP_RESTRICT_ALL, "all"},
    /* Compatibility values (do we want these?) */
    { CLIP_RESTRICT_ALL, "true"},
    { CLIP_RESTRICT_ALL, "on"},
    { CLIP_RESTRICT_ALL, "yes"},
    BITMASK_STRING_END_OF_LIST
};

. . .
     char unrecognised[256];
       if (g_strcasecmp(name, "RestrictOutboundClipboard") == 0)
       {
            cfg->restrict_outbound_clipboard = g_str_to_bitmask(value, restrict_map, ',', unrecognised, sizeof(unrecognised));
            if (unrecognised[0] != '\0')
            {
                LOG(LOG_LEVEL_WARNING,
                         "Unrecognised tokens parsing 'RestrictOutboundClipboard' %s",
                        unrecognised);
            }
        }
```

I think this is more likely to be useful for parsing other settings in files. It also means that adding an extra value just needs a line adding to the bitmask, and an extra line adding to `restrict_map` so it increases maintainability.

Also, (and this is important but we haven't covered it yet) we can add some test cases for this function to the test suite. Don't worry about this yet - let's get the basics in first.

The implementation of `g_str_to_bitmask()` can make a single pass through the string. You'll need to cope with whitespace in the string somehow. There's a function called `g_strtrim()` in the string library which can help you with this. Use `LOG(LOG_LEVEL_WARNING, 

Please come back to me if you think this isn't a reasonable way to proceed, or if I haven't explained myself very clearly. Also, there may well be better names to use than some of the ones I've picked above. I'm just trying to give you an idea rather than write your code for you.

Reply:
Thanks, I've added `g_str_to_bitmask` and changed to use it.